### PR TITLE
Do not keep task id reference indefinitely in Celery instrumenation

### PR DESCRIPTION
**Pull request by rbagd**
# Description

Reference to a task identifier as a string is kept indefinitely by the instrumentation. This just makes sure it gets removed once no longer needed.

Related #3458

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [X] Unit test to ensure reference is not there

# Does This PR Require a Core Repo Change?

- [X] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [X] Followed the style guidelines of this project
- [X] Changelogs have been updated
- [X] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-celery/tests/test_metrics.py [hunk 1]
```diff
@@ -42,6 +42,8 @@ def test_basic_metric(self):
         task_runtime_estimated = (default_timer() - start_time) * 1000
 
         metrics = self.get_metrics()
+        self.assertEqual(CeleryInstrumentor().task_id_to_start_time, {})
```
Comment (rads-1996) [inline instrumentation/opentelemetry-instrumentation-celery/tests/test_metrics.py:45]: Should another test be added where the task_id does not exist and the pop operation returns None instead?

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- `opentelemetry-instrumentation-celery`: Fix a memory leak where a reference to a task identifier is kept indefinitely
+  ([#3463](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3463))
```
Comment (xrmx) [inline CHANGELOG.md]:
```suggestion
  ([#3690](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3690))
```

File: instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py [hunk 3]
```diff
@@ -362,6 +362,7 @@ def _record_histograms(self, task_id, metric_attributes):
             self.task_id_to_start_time.get(task_id),
             attributes=metric_attributes,
         )
+        self.task_id_to_start_time.pop(task_id, None)
```
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py:365]: What is that we are leaking? On line 355 we are setting this a time object so don't expect to keep alive something on the celery side?

Comment (rbagd) [reply instrumentation/opentelemetry-instrumentation-celery/src/opentelemetry/instrumentation/celery/__init__.py:365]: `task_id_to_start_time` is a dictionary initialised with `CeleryInstrumentor`. Each time a task finishes the dictionary is augmented with a record of time spent for that task id. In the current state the dictionary just keeps growing even though the task finished long time ago (task identifiers are normally unique). There are no references to any other Python objects that prevent GC, it's just this relatively small issue.

File: instrumentation/opentelemetry-instrumentation-celery/tests/test_metrics.py (hunk 1 above)
Comment (rbagd) [reply instrumentation/opentelemetry-instrumentation-celery/tests/test_metrics.py:45]: Would that test matter actually? We're not testing return value of the `pop` here in any case. I may not see your point. Can you maybe explain the case you have in mind?
